pkg/objstore/client/factory.go

```diff
@@ -3,39 +3,57 @@ package client
 import (
 	"context"
 	"fmt"
-	"runtime"
 
-	"cloud.google.com/go/storage"
 	"github.com/go-kit/kit/log"
 	"github.com/improbable-eng/thanos/pkg/objstore"
 	"github.com/improbable-eng/thanos/pkg/objstore/gcs"
 	"github.com/improbable-eng/thanos/pkg/objstore/s3"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
-	"github.com/prometheus/common/version"
-	"google.golang.org/api/option"
+	yaml "gopkg.in/yaml.v2"
 )
 
-var ErrNotFound = errors.New("no valid GCS or S3 configuration supplied")
+type objProvider string
+
+const (
+	GCS objProvider = "GCS"
+	S3  objProvider = "S3"
+)
+
+type BucketConfig struct {
+	Type   objProvider `yaml:"type"`
+	Config interface{} `yaml:"config"`
+}
+
+var ErrNotFound = errors.New("not found bucket")
 
 // NewBucket initializes and returns new object storage clients.
-func NewBucket(logger log.Logger, gcsBucket *string, s3Config s3.Config, reg *prometheus.Registry, component string) (objstore.Bucket, error) {
-	if *gcsBucket != "" {
-		gcsOptions := option.WithUserAgent(fmt.Sprintf("thanos-%s/%s (%s)", component, version.Version, runtime.Version()))
-		gcsClient, err := storage.NewClient(context.Background(), gcsOptions)
-		if err != nil {
-			return nil, errors.Wrap(err, "create GCS client")
-		}
-		return objstore.BucketWithMetrics(*gcsBucket, gcs.NewBucket(*gcsBucket, gcsClient, reg), reg), nil
+func NewBucket(logger log.Logger, conf string, reg *prometheus.Registry, component string) (objstore.Bucket, error) {
```

Comment:
yup, let's do it

Reply:
if you are afraid in multilines you can always do json and minimize it ;p

Reply:
This flag is just for bucket configuration, it would work or not, nothing to debug in here. But I imagine seeing the issues of putting required indentations inside a multiline config to pass through the argument.

For the consistency, I would move options to the config file as well, but this is another question.

Reply:
agree with @Bplotka , let's try to do it. Maybe we can provide a way to generate the bucket configuration in yaml format contained in `--objstore.config`? Also user can use the `thanos bucket ls` to verify the flag for bucket configuration.